src/mempoollimit.cpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2019 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php 
+
+#include "core_memusage.h"
+#include "mempoollimit.h"
+#include "random.h"
+#include "serialize.h"
+#include "timedata.h"
+#include "version.h"
+
+const CAmount DEFAULT_FEE = 10000;
+const TxWeight ZERO_WEIGHT = TxWeight(0, 0);
+
+void RecentlyEvictedList::pruneList()
+{
+    if (txIdSet.empty()) {
+        return;
+    }
+    int64_t now = GetAdjustedTime();
+    size_t startIndex = (txIdsAndTimesIndex + capacity - txIdSet.size()) % capacity;
+    boost::optional<std::pair<uint256, int64_t>> txIdAndTime;
+    while ((txIdAndTime = txIdsAndTimes[startIndex]).is_initialized() && (now - txIdAndTime.get().second) > timeToKeep) {
+        txIdsAndTimes[startIndex] = boost::none;
+        txIdSet.erase(txIdAndTime.get().first);
+        startIndex = (startIndex + 1) % capacity;
+    }
+}
+
+void RecentlyEvictedList::add(const uint256& txId)
+{
+    pruneList();
+    if (txIdsAndTimes[txIdsAndTimesIndex].is_initialized()) {
+        auto txIdAndTime = txIdsAndTimes[txIdsAndTimesIndex];
+        txIdSet.erase(txIdAndTime.get().first);
```

Comment:
So.. we just overwrite anything existing in there? This seems like someone could fill up the RecentlyEvictedList with permutations of (malleable) transactions and we still end up in DoS. What have I missed?

Reply:
Note that the `RecentlyEvictedList` is not intended to mitigate intentional DoS.

Reply:
We could make the size of the recently evicted list configurable, but I am not sure about how beneficial that would be.